src/main/utility/counter.rs

```diff
@@ -0,0 +1,178 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+/*!
+A counter that can be used to count frequencies of a set of objects. The counter starts
+with no keys. When an unknown key is incremented, the counter adds a new key to an
+internal map and sets the count for that key to 1. When a known key is incremented, the
+count for that key increases. The state of the counter can be extracted by converting it
+to a string, which lists the counts for all keys sorted with the heaviest hitters first.
+Currently, only String types are supported, but we may eventually support counting
+generic types.
+*/
+
+use std::collections::btree_map::BTreeMap;
+use std::iter::FromIterator;
+
+/// The main counter object that maps individual keys to count values.
+pub struct Counter {
+    // TODO: convert this so we could count generic types instead of Strings
+    items: BTreeMap<String, u64>,
+}
+
+impl Counter {
+    /// Initializes a new counter map that starts with no keys.
+    fn new() -> Counter {
+        Counter {
+            items: BTreeMap::new(),
+        }
+    }
+
+    /// Increment the counter value for the key given by id.
+    /// Returns the value of the counter after it was incremented.
+    fn add_one(&mut self, id: &str) -> u64 {
+        let count = self.items.entry(id.to_string()).or_insert(0);
+        *count = *count + 1;
+        *count
+    }
+
+    /// Returns the counter value for the key given by id, or 0 if the key has not
+    /// previously been incremented.
+    // Used for unit test
+    #[allow(dead_code)]
+    fn get_value(&mut self, id: &str) -> u64 {
+        match self.items.get(&id.to_string()) {
```

Comment:
Unless I'm mistaken, you should be able to pass `id` here without converting to `String`.

Reply:
Done.